quixstreams/state/rocksdb/transaction.py

```diff
@@ -295,72 +326,149 @@ def failed(self) -> bool:
-            for k, v in self._update_cache[cf_name].items():
-                self._partition.produce_to_changelog(
-                    key=k, value=v if v is not _deleted else None, headers=headers
-                )
-                offset += 1
+    @property
+    def changelog_topic_partition(self) -> Optional[Tuple[str, int]]:
+        """
+        Return the changelog topic-partition for the StorePartition of this transaction.
 
-        self._batch.put(
-            CHANGELOG_OFFSET_KEY, int_to_int64_bytes(offset), meta_cf_handle
-        )
-        logger.debug(f"Changelog offset set to {offset}")
+        Returns `None` if changelog_producer is not provided.
 
-    @_validate_transaction_state
-    def maybe_flush(self, offset: Optional[int] = None):
+        :return: (topic, partition) or None
         """
-        Flush the recent updates to the database and empty the update cache.
-        It writes the WriteBatch to RocksDB and marks itself as finished.
+        if self._changelog_producer is not None:
+            return (
+                self._changelog_producer.changelog_name,
+                self._changelog_producer.partition,
+            )
 
-        If writing fails, the transaction will be also marked as "failed" and
-        cannot be used anymore.
+    def as_state(self, prefix: Any = DEFAULT_PREFIX) -> TransactionState:
+        """
+        Create a one-time use `TransactionState` object with a limited CRUD interface
+        to be provided to `StreamingDataFrame` operations.
 
-        >***NOTE:*** If no keys have been modified during the transaction
-            (i.e. no "set" or "delete" have been called at least once), it will
-            not flush ANY data to the database including the offset in order to optimize
-            I/O.
+        The `TransactionState` will prefix all the keys with the supplied `prefix`
+        for all underlying operations.
 
-        :param offset: offset of the last processed message, optional.
+        :param prefix: a prefix to be used for all keys
+        :return: an instance of `TransactionState`
         """
-        try:
-            # Don't write batches if this transaction doesn't change any keys
-            if len(self._batch):
-                cf_handle = self._partition.get_column_family_handle(METADATA_CF_NAME)
-                if offset is not None:
-                    self._batch.put(
-                        PROCESSED_OFFSET_KEY, int_to_int64_bytes(offset), cf_handle
+        return TransactionState(
+            transaction=self,
+            prefix=(
+                prefix
+                if isinstance(prefix, bytes)
+                else serialize(prefix, dumps=self._dumps)
+            ),
+        )
+
+    def _produce_changelog(self, processed_offset: Optional[int] = None):
+        changelog_producer = self._changelog_producer
+        if changelog_producer is None:
+            return
+
+        source_topic, changelog_topic, partition = (
```

Comment:
`source_topic` no longer used

Reply:
Agree, I removed it from the changelog classes